ingestion/moj_statistical_publications_source/source.py

```diff
@@ -0,0 +1,298 @@
+import logging
+from io import BufferedReader
+from typing import Any, Dict, Generator, Iterable, List, Optional
+from urllib.parse import urljoin
+
+import datahub.emitter.mce_builder as mce_builder
+import datahub.emitter.mcp_builder as mcp_builder
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.api.decorators import (
+    SupportStatus,
+    config_class,
+    platform_name,
+    support_status,
+)
+from datahub.ingestion.api.source import (
+    CapabilityReport,
+    SourceReport,
+    TestableSource,
+    TestConnectionReport,
+)
+from datahub.ingestion.api.workunit import MetadataWorkUnit
+from datahub.ingestion.source.common.subtypes import DatasetContainerSubTypes
+from datahub.metadata.com.linkedin.pegasus2avro.common import TimeStamp
+from datahub.metadata.schema_classes import (
+    ContainerClass,
+    DataPlatformInstanceClass,
+    DatasetPropertiesClass,
+    DomainsClass,
+    GlobalTagsClass,
+    SubTypesClass,
+    TagAssociationClass,
+)
+from datahub.utilities.time import datetime_to_ts_millis
+
+from ingestion.ingestion_utils import FindMojDataEntityTypes, format_domain_name
+
+from .api_client import MojPublicationsAPIClient
+from .config import MojPublicationsAPIConfig
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+@platform_name("File")
+@config_class(MojPublicationsAPIConfig)
+@support_status(SupportStatus.CERTIFIED)
+class MojPublicationsAPISource(TestableSource):
+    """
+    This plugin pulls metadata from the gov.uk publications search and content APIs
+    """
+
+    def __init__(
+        self,
+        ctx: PipelineContext,
+        config: MojPublicationsAPIConfig,
+        validate_domains: bool = True,
+    ) -> None:
+        self.ctx = ctx
+        self.config = config
+        self.report = SourceReport()
+        self.fp: Optional[BufferedReader] = None
+        self.client = MojPublicationsAPIClient(
+            config.base_url, config.default_contact_email, config.params
+        )
+        if validate_domains:
+            self.client.validate_domains()
+        self._id_to_domain_contact_mapping = self.client._id_to_domain_contact_mapping
+        self.platform_name = "GOV.UK"
+        self.platform_instance = "ministry-of-justice-publications"
+        self.access_requirements = config.access_requirements
+        self.web_url = self.config.base_url.removesuffix("/api").removesuffix("/api/")
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = MojPublicationsAPIConfig.model_validate(config_dict)
+        return cls(ctx, config)
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        all_publications_metadata = self.client.list_all_publications_metadata()
+        collections_metadata = self.client.get_collections_from_all_results(
+            all_publications_metadata, self.config.collections_to_exclude
+        )
+
+        mcps = []
```

Comment:
yeah, this might be left over from the start when i thought i might be adding more aspects from different methods but i did it all in one so i'll remove